generator/src/main/java/com/graphicsfuzz/generator/mutateapi/ExpressionGenerator.java

```diff
@@ -0,0 +1,298 @@
+          return generateFunction(factManager,
+              functionDefinition,
+              stmt,
+              value,
+              generator);
+        case 2:
+          // Retrieving and return the existing fact from the Fact manager
+          final Optional<Expr> expr = factManager.getFact(value);
+          if (expr.isPresent()) {
+            return expr.get();
+          } else {
+            continue;
+          }
+          // Generate ConstantExpr (Float, Int)
+        case 3:
+          return generateLiteralNumber(value);
+        default:
+          return null;
+      }
+    }
+  }
+
+  // TODO: Handle the cases of vectors (vec4, vec3, vec2).
+  private String genVarName(Value value) {
+    return Constants.GLF_PRIMITIVE_VAR + "_"
+        + value.getType().toString() + "_"
+        + value.getData().get(0).get().toString().replace(".", "_");
+  }
+
+  // TODO: Requires further implementation in the case of the negative number, i.e., -0.75 could be
+  // converted to _GLF_FLOAT_NEGATIVE_0_75.
+  private String nameFromNumber(Number number) {
+    return number.toString();
+  }
+
+  private String getFunctionName(Value value) {
+    return Constants.GLF_COMPUTE + "_"
+        + value.getType().toString().toUpperCase() + "_"
+        + nameFromNumber(value.getData().get(0).get()).replace(".", "_")
+        + "_ID_" + idGenerator.freshId();
+  }
+
+  // TODO: Handles the case of boolean value.
+  private Optional<Number> numberFromName(String string) {
+    try {
+      return Optional.of(NumberFormat.getInstance().parse(string));
+    } catch (ParseException parseException) {
+      return Optional.empty();
+    }
+  }
+
+  private Expr generateVariableFact(FactManager factManager,
+                                    FunctionDefinition functionDefinition,
+                                    Stmt stmt,
+                                    Value value,
+                                    IRandom generator) {
+    final String variableName = genVarName(value);
+    final VariableDeclInfo declInfo =
+        new VariableDeclInfo(variableName, null,
+            new Initializer(new FloatConstantExpr(value.getData().get(0).get().toString())));
+    final VariablesDeclaration variablesDecl = new VariablesDeclaration(value.getType(), declInfo);
+    functionDefinition.getBody().addStmt(new DeclarationStmt(variablesDecl));
+
+    // Store variable fact in FactManager
+    factManager.addVariableFact(declInfo, new VariableFact(variablesDecl, declInfo, value));
+    return new VariableIdentifierExpr(variableName);
+  }
+
+  private Expr generateFunction(FactManager factManager,
+                                FunctionDefinition functionDefinition,
+                                Stmt stmt,
+                                Value targetValue,
+                                IRandom generator) {
+
+    final String functionName = getFunctionName(targetValue);
+    final FactManager childFactManager = factManager.clone();
```

Comment:
^ Rather than cloning the fact manager, I suppose we would really like a new fact manager that has global scope facts available to it.

Perhaps give FactManager a method, "getGlobalFacts", which could return this if the fact manager has no prototype, and could return prototype.getGlobalFacts() otherwise.  That should give the facts at global scope.  Then the child fact manager could be a clone of factManager.getGlobalFacts().